### PR TITLE
fix(ValidateHostHandler): add null and empty host checks, fix IPv6  Bracket Stripping

### LINKED ISSUE #7746: `ValidateHostHandler`  Off-by-One in IPv6 Bracket Stripping & Missing Null-Check on Host Header
`ValidateHostHandler` contains two related core server bugs in the `isValidHost()` method:
1. **Off-by-one error** when stripping brackets from IPv6 addresses causes the bare IPv6 address to be truncated by one character, making any custom non-loopback IPv6 hostname set via `refine.host` **always rejected**, even when it is valid. (Line 81: `host.substring(1, host.length() - 2)`)
2. **Missing null-check** on the `Host` request header. HTTP/1.0 clients (and some reverse proxies) may omit the `Host` header. When that happens, `isValidHost(null)` is invoked, which throws a `NullPointerException` (Line 92).

### To Reproduce
Steps to reproduce the behavior for **Bug 1 (IPv6 off-by-one)**:
1. Start OpenRefine with a custom non-loopback IPv6 host (e.g., `./refine -H 2001:db8::1`).
2. Make an HTTP request to the server with `curl -H "Host: [2001:db8::1]:3333" http://localhost:3333/`.
3. The server immediately rejects the request.

Steps to reproduce the behavior for **Bug 2 (Null Host Header NPE)**:
1. Start OpenRefine normally.
2. Send an HTTP/1.0 request without a `Host` header: `printf 'GET / HTTP/1.0\r\n\r\n' | nc localhost 3333`.
3. The server crashes with an unhandled exception instead of properly rejecting the invalid request.

### Current Results
- **For Bug 1:** Valid IPv6 requests yield a `404 Not Found — Invalid hostname` because the string `[2001:db8::1]` is incorrectly stripped to `"2001:db8::"` (missing the trailing `1`), failing the validation check.
- **For Bug 2:** An unhandled `HTTP 500 Internal Server Error` is thrown due to a `NullPointerException` propagating from `LOOPBACK_PATTERN.matcher(null).find()`.

### Expected Behavior
- **For Bug 1:** The server should accurately parse the IPv6 address from the `Host` header and accept it (i.e. return `200 OK`). To achieve this, the bracket stripping logic should use `length() - 1` instead of `length() - 2`.
- **For Bug 2:** The server should cleanly return a `404 Not Found — Invalid hostname` rather than throwing a `NullPointerException` when a `Host` header is missing. A simple `if (host == null)` guard is needed.

### Versions
 - Operating System: All (OS-independent core server logic)
 - Browser Version: N/A
 - JRE or JDK Version: All supported Java versions
 - OpenRefine: Master / Current Development Branch

### Datasets
N/A - This is a server code bug, independent of user datasets.

### Additional context
**Affected File:** `server/src/com/google/refine/ValidateHostHandler.java`

**Diff to fix both issues:**
```diff
--- a/server/src/com/google/refine/ValidateHostHandler.java
+++ b/server/src/com/google/refine/ValidateHostHandler.java
@@ -65,17 +65,21 @@ class ValidateHostHandler extends HandlerWrapper {
     public boolean isValidHost(String host) {
 
+        // Reject missing Host header immediately
+        if (host == null || host.isEmpty()) {
+            return false;
+        }
+
         // Allow loopback IPv4 and IPv6 addresses, as well as localhost
         if (LOOPBACK_PATTERN.matcher(host).find()) {
             return true;
         }
 
         // Strip port from hostname - for IPv6 addresses, if
         // they end with a bracket, then there is no port
         int index = host.lastIndexOf(':');
         if (index > 0 && !host.endsWith("]")) {
             host = host.substring(0, index);
         }
 
         // Strip brackets from IPv6 addresses
         if (host.startsWith("[") && host.endsWith("]")) {
-            host = host.substring(1, host.length() - 2);
+            host = host.substring(1, host.length() - 1);
         }
 
         // Allow only if stripped hostname matches expected hostname
         return expectedHost.equalsIgnoreCase(host);
     }
```

*Note: With comprehensive unit tests added to verify these fixes against valid IPv6 addresses, null hosts, and empty hosts.*

### PR DESCRIPTION
**Pull request by Manishnemade12**
# Pull Request: Fix ValidateHostHandler IPv6 and Null Host Bugs

Fixes: #7746 

## Summary
This PR fixes two critical bugs in the `ValidateHostHandler` class:

1. **IPv6 Bracket Off-by-One Error:**
   - The previous logic incorrectly stripped the last character from IPv6 addresses in the Host header, causing valid IPv6 hosts to be rejected.
   - Fixed by changing `host.substring(1, host.length() - 2)` to `host.substring(1, host.length() - 1)`.

2. **Null Host Header Handling:**
   - The code did not check for null or empty Host headers, resulting in a `NullPointerException` for HTTP/1.0 requests or some reverse proxies.
   - Fixed by adding an explicit null/empty check at the start of `isValidHost()`.

## Additional Improvements
- Added comprehensive unit tests for:
  - Valid IPv6 host acceptance
  - Null and empty Host header rejection
  - Handler returning 404 for missing Host header
- Ensured test dependencies (`testng`, `mockito`) are present and test source directory is configured in `server/pom.xml`.

## Files Changed
- `server/src/com/google/refine/ValidateHostHandler.java`: Bugfixes for IPv6 and null Host handling.
- `server/tests/src/com/google/refine/ValidateHostHandlerTest.java`: New unit tests for all edge cases.
- `server/pom.xml`: Ensures test dependencies and test source directory are configured.

## How to Test
- Run `mvn -pl server -am test -Dtest=ValidateHostHandlerTest` to verify all tests pass.
- Manual validation:
  - Start OpenRefine with a custom IPv6 host and test with a valid Host header.
  - Send HTTP/1.0 requests without a Host header and confirm a 404 is returned, not a 500.

---

**This PR ensures OpenRefine is robust against Host header parsing errors and DNS rebinding attacks.**

### CODE REVIEW

**Comment by Manishnemade12** [comment]
@tfmorris Can you please review this pr .

**Review by tfmorris — CHANGES_REQUESTED** [review]
Thanks for the contribution. Looks good, but please address Copilot's review comments and the formatting issue. It should be ready to merge after that.

You can use the following command to invoke the formatter:
```
./refine lint
```

File: server/tests/src/com/google/refine/ValidateHostHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package com.google.refine;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
```
[the remark this answers is not in the document]
Comment (Manishnemade12) [reply server/tests/src/com/google/refine/ValidateHostHandlerTest.java:37]: done

**Comment by Manishnemade12** [comment]
@tfmorris Added OpenRefine BSD license header to keep licensing consistent with other Java sources in the project.
Please review it once

File: server/tests/src/com/google/refine/ValidateHostHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
+/*
+
+Copyright 2010, Google Inc.
```
Comment (tfmorris) [inline server/tests/src/com/google/refine/ValidateHostHandlerTest.java]:
```suggestion
Copyright 2026, OpenRefine contributors
```

File: server/tests/src/com/google/refine/ValidateHostHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        HttpServletRequest request = mock(HttpServletRequest.class);
+        HttpServletResponse response = mock(HttpServletResponse.class);
+
+        when(request.getHeader("Host")).thenReturn(null);
+
+        handler.handle("/", baseRequest, request, response);
+
+        verify(response).sendError(HttpServletResponse.SC_NOT_FOUND, "Invalid hostname");
+    }
+}
```
Comment (tfmorris) [inline server/tests/src/com/google/refine/ValidateHostHandlerTest.java]:
```suggestion
}

```

**Review by tfmorris — CHANGES_REQUESTED** [review]
Thanks for adding the module header. I've updated it with the current copyright info.

The tests are failing and I suspect that you didn't run:

```
./refine lint
```

Please make sure that all tests are passing.

**Comment by Manishnemade12** [comment]
@tfmorris 
image
i did it .

**Comment by Manishnemade12** [comment]
@tfmorris i addressed your all comments can you please review this pr.

**Comment by Manishnemade12** [comment]
@tfmorris i addressed your all comments can you please review this pr.

**Review by tfmorris — APPROVED** [review]
Perfect! Thanks very much. Sorry for the delay in reviewing.
